lib/engine/game/g_1849/step/buy_token.rb

```diff
@@ -0,0 +1,154 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/base'
+
+module Engine
+  module Game
+    module G1849
+      module Step
+        class BuyToken < Engine::Step::Base
+          ACTIONS = %w[buy_token pass].freeze
+          MIN_PRICE = 1
+
+          def actions(entity)
+            return [] unless entity == current_entity
+            return [] unless can_buy_token?(entity)
+
+            ACTIONS
+          end
+
+          def round_state
+            super.merge(
+              {
+                bought_token: false,
+              }
+            )
+          end
+
+          def setup
+            super
+            @round.bought_token = false
+          end
+
+          def can_buy_token?(entity)
+            @game.acquiring_station_tokens? &&
+              @game.phase.tiles.include?(:green) &&
+              current_entity == entity &&
+              !@round.bought_token &&
+              !available_tokens(entity).empty? &&
+              MIN_PRICE <= buying_power(entity)
+          end
+
+          def available_tokens(entity)
+            entity.tokens_by_type
+          end
+
+          def can_sell_token?(token, entity)
+            token.corporation.placed_tokens.size > 1 &&
+            token != token.corporation.tokens.first &&
+            !token.city.tokened_by?(entity)
+          end
+
+          # look for any cities reachable from entity that are tokened by another corporation that
+          # has at least 2 tokens placed
+          #
+          # this is expensive - move to auto_actions
+          def any_buyable_tokens_placed?(entity)
+            @game.token_graph_for_entity(entity).connected_nodes(entity).keys.each do |node|
+              next unless node.city?
+
+              node.tokens.each do |token|
+                next unless token
+                next if token.corporation == entity
+
+                return true if can_sell_token?(token, entity)
+              end
+            end
+            false
+          end
+
+          def auto_actions(entity)
+            return [Engine::Action::Pass.new(entity)] unless any_buyable_tokens_placed?(entity)
+
+            super
+          end
+
+          # 1849 doesn't allow more than one token per corporation per tile
+          def can_token_city?(entity, city)
+            city.tile.nodes.select(&:city?).none? { |c| c.tokened_by?(entity) }
+          end
+
+          def can_replace_token?(entity, token)
```

Comment:
How is this different from `can_sell_token?`

Reply:
If this is related to buying a token, why is it called `can_replace_token?` instead of `can_buy_token?`?